app/models/mixins/purging_mixin.rb

```diff
@@ -200,24 +200,33 @@ def purge_in_batches(conditions, window, total = 0, total_limit = nil)
           query = query.limit(current_window)
         end
 
-        if respond_to?(:purge_associated_records)
-          # pull back ids - will slow performance
-          batch_ids = query.pluck(:id)
-          break if batch_ids.empty?
-
-          current_window = batch_ids.size
-        else
-          batch_ids = query
+        count = 0
+
+        # Purge each batch with their associated records
+        # in a transaction to avoid leaving orphaned references.
+        transaction do
+          if respond_to?(:purge_associated_records)
+            # pull back ids - will slow performance
+            batch_ids = query.pluck(:id)
+            current_window = batch_ids.size
+          else
+            batch_ids = query
+          end
+
+          unless batch_ids.empty?
```

Comment:
convert a early break to a simple conditional

Reply:
This might be a problem..in the case of the else condition, we don't want it to execute the query, but instead stay as a scope.

Reply:
Ah, good eye.  I wanted to test that else condition.